### PR TITLE
Fix: check if target is valid

### PR DESCRIPTION
**Pull request by mtojek**
This PR fixes the issue reported in https://github.com/elastic/go-ucfg/issues/159 .

Check if the target is valid, before setting the value. The target field will stay `nil`.

/cc @ruflin

### CODE REVIEW

File: reify.go [hunk 1]
```diff
@@ -343,7 +343,9 @@ func reifyGetField(
 		return err
 	}
 
-	to.Set(pointerize(to.Type(), v.Type(), v))
+	if v.IsValid() {
+		to.Set(pointerize(to.Type(), v.Type(), v))
+	}
```
Comment (urso) [inline reify.go:348]: Given the test case I don't see why `v` can be invalid. Do we have an original issue with sample code? `v` should always be assignable to `to`, and return a zero value at least.

As reifyGetField is an internal function processing a sub-tree only, I wonder if the issue can occur in other edge cases as well.

File: yaml/yaml_test.go [hunk 2]
```diff
@@ -113,3 +113,38 @@ func TestArray(t *testing.T) {
...
+	}
+
+	var verify hasVars
+	err = c.Unpack(&verify)
+	assert.Nil(t, err)
+
+	assert.Nil(t, verify.Vars[0].Default)
+	assert.Nil(t, verify.Vars[1].Default)
+	assert.Equal(t, uint64(3), verify.Vars[2].Default)
+	assert.Len(t, verify.Vars[3].Default.([]interface{}), 3)
```
Comment (urso) [inline yaml/yaml_test.go:149]: Can we split this up into 4 test cases (e.g. table driven test), or is the error only triggered by this setup?

Was the error triggered for each of these cases?

File: yaml/yaml_test.go [hunk 3]
```diff
@@ -113,3 +113,38 @@ func TestArray(t *testing.T) {
...
 	assert.Equal(t, verify[1]["c"], 4)
 }
+
+type hasVars struct {
+	Vars []variable
+}
+
+type variable struct {
+	Default interface{}
+}
```
Comment (urso) [inline yaml/yaml_test.go:123]: these types are unique to the test case only. Consider moving the type declaration into TestStructEmptyArray

File: yaml/yaml_test.go (hunk 2 above)
Comment (urso) [reply yaml/yaml_test.go:149]: Let's add a sample test with an empty array as has been shown in the issue.

**Comment by urso** [comment]
Is this error reproducible if we just unpack into an array, instead of an array within a struct?

File: reify.go (hunk 1 above)
Comment (mtojek) [reply reify.go:348]: > Do we have an original issue with sample code?

This is the exact extracted part from the linked PR in the issue. Just unpack the given YAML file into a structure with `interface{}`

File: yaml/yaml_test.go (hunk 2 above)
Comment (mtojek) [reply yaml/yaml_test.go]: Good catch. It seems that an empty array is also an issue. I simplified the test case.

File: yaml/yaml_test.go (hunk 3 above)
Comment (mtojek) [reply yaml/yaml_test.go]: Done.

**Comment by mtojek** [comment]
> Is this error reproducible if we just unpack into an array, instead of an array within a struct?

I replied above. It's reproducible - I simplified the test case, a single line seems to be enough.

**Comment by ruflin** [comment]
Should we add a chanelog entry?

File: reify.go (hunk 1 above)
Comment (urso) [reply reify.go:348]: What I'm wondering about is, if this or a similar error can occur when we unpack into some alternative type. For example if one unpacks into a `map`, then the call `to.SetMapIndex(key, v)` at line 224 would be the equivalent for maps. For arrays/slices the equivalent call is `to.Index(idx).Set(v)` in `reifyDoArray`. Both calls do not check for an invalid `reflect.Value` yet. Given the bug at hand, I doubt the other 2 cases have tests that might show an error.

**Comment by urso** [comment]
> Should we add a chanelog entry?

Yes, please.

**Comment by mtojek** [comment]
> > Should we add a chanelog entry?
> 
> Yes, please.

Fixed.

File: reify.go (hunk 1 above)
Comment (mtojek) [reply reify.go:348]: I've added more test cases, but couldn't reproduce. If you see any potential weird structure that can cause an error, I can add a test case.

Other way round is to wrap all `to.Set()` calls with a condition checking if the target is valid.

**Comment by mtojek** [comment]
I think we might have hit also this: https://github.com/golang/go/commit/201cb046b745f8bb00e3d382290190c74ba7b7e1

File: error_test.go [hunk 4]
```diff
@@ -218,7 +219,18 @@ func TestErrorMessages(t *testing.T) {
...
+
+// adjustMessageFormat method modifies the message to be backward compatible with previous Go releases.
+// The change introduced in https://github.com/golang/go/commit/201cb046b745f8bb00e3d382290190c74ba7b7e1 added
+// quotas and makes some test assertions failing.
+func adjustMessageFormat(message string) string {
+	adjusted := strings.Replace(message, "unknown unit \"", "unknown unit ", 1)
+	adjusted = strings.Replace(adjusted, "\" in duration \"", " in duration ", 1)
+	adjusted = strings.Replace(adjusted, "\" accessing", " accessing", 1)
+	return adjusted
+}
```
Comment (urso) [inline error_test.go:236]: Which go version failed here? For go `tip` only? If we need it for tip and it is a known error in `tip` to be fixed soon I would ignore it for now. We check with `tip` to see if we might be hit by a change in reflection in the stdlib early in time.

If it fails for a released go version let's add a conditional.

File: reify.go (hunk 1 above)
Comment (urso) [reply reify.go:348]: If the tests show that we are fine, I'm ok with that.

maybe we can add these combinations for maps as well:
```
map[string]interface{}{}  , `{"a": []}`
map[string]interface{}{"a": nil}  , `{"a": []}`
map[string][]interface{}{}, `{"a": []}`
map[string][]interface{}{"a": nil}, `{"a": []}`
```

and for arrays:

```
[]interface{} , `[[]]`
[][]interface{}, `[[]]`
```

File: yaml/yaml_test.go [hunk 5]
```diff
@@ -113,3 +113,119 @@ func TestArray(t *testing.T) {
...
+	c, err := NewConfig(input)
+	if err != nil {
+		t.Fatalf("failed to parse input: %v", err)
+	}
+
+	var verify map[string]interface{}
+	err = c.Unpack(&verify)
+	assert.Nil(t, err)
+	assert.Equal(t, map[string]interface{}{}, verify)
+}
```
Comment (urso) [inline yaml/yaml_test.go:231]: there is quite some repetition here. How about moving the test cases into a table like so?

```
cases := map[string]struct{
  input string
  to interface{}
  want interface{}
}{
  "empty map into map of interfaces": {
    input: "{}",
    to: &map[string]interface{}{},
    want: &map[string]interface{}{},
  },
  ...
}

for name, test := range cases {
  t.Run(name, func(t *testing.T) {
    c, err := MustNewConfig(byte(test.input))
    mustUnpack(t, c, test.to)
    assert.Equal(t, test.want, test.to)
  })
}
```

File: reify.go (hunk 1 above)
Comment (mtojek) [reply reify.go:348]: I've updated test cases, but I think we need to review/accept the final unmarshaled forms (I'm referring specifically to this case: `map[string]interface{}{"a": nil}` where the field disappears).

File: yaml/yaml_test.go (hunk 5 above)
Comment (mtojek) [reply yaml/yaml_test.go:242]: Fixed.

File: reify.go (hunk 1 above)
Comment (urso) [reply reify.go:348]: Hm, the field `a` should not just disappear. If the input configuration does not contain the field, or if it is an empty array, then it should not be overwritten. We want to merge the configuration into the object given. Yet one can discuss what it means to merge an empty array... But it kinda looks like we just found another bug.
I don't want increase the scope for this PR any further. How about skipping this tests that do remove `a` and create a follow up issue?

File: yaml/yaml_test.go (hunk 5 above)
Comment (urso) [reply yaml/yaml_test.go:242]: `mustUnpack` should be available already if you rebase.

File: error_test.go [hunk 6]
```diff
@@ -218,7 +220,21 @@ func TestErrorMessages(t *testing.T) {
 			}
 
 			golden := string(tmp)
-			assert.Equal(t, golden, message)
+			message = adjustMessageFormat(message)
+			assert.Equal(t, golden, message, "Go runtime version: %s", runtime.Version())
```
Comment (urso) [inline error_test.go:224]: 💯

File: error_test.go (hunk 4 above)
Comment (urso) [reply error_test.go:240]: Oh, it's a bug in go1.14 :(

Checking the travis file, I noticed 1.14 is still missing.

Comment (mtojek) [reply error_test.go:241]: In this case it was "devel". I can adjust the Travis file too, unless you think this should go in a separate PR.

File: yaml/yaml_test.go (hunk 5 above)
Comment (mtojek) [reply yaml/yaml_test.go:236]: Rebased and refactored as it wasn't exposed.

File: reify.go (hunk 1 above)
Comment (mtojek) [reply reify.go:355]: I think I've just fixed this case by checking if the target is valid:

```go
if v.IsValid() {
    to.SetMapIndex(key, v)
}
```

File: error_test.go (hunk 4 above)
Comment (urso) [reply error_test.go:241]: Hm, if it's just tip I'd remove the function altogether. Although I'd prefer to add go 1.14 in another PR, I'd like to have some assurance that 1.14 does not contain the bug :)

Comment (mtojek) [reply error_test.go:241]: Let me prepare a PR for this.
